### PR TITLE
Feature to parse the XML doctype and add it back to XML output

### PR DESCRIPTION
**Pull request by gnott**
https://elifesciences.atlassian.net/browse/ELPP-2379   The new XML kitchen sink has a doctype denoting JATS v1.1.  Before, xmlio was always adding a doctype for JATS v1.1d3.

Added here: to parse() you can optionally return the doctype from ElementTree.XMLParser using a custom parser. 

In output(), you can supply a dict of values to be used in the doctype when the XML file is generated.

It should be backwards compatible while making it possible to use this new ability. Most specifically, I did not like how in elife-bot when we rewrite the xlink:href values when applying the version number, it would alter the doctype of the XML from what was read in to something different.

### CODE REVIEW

**Comment by gnott** [comment]
@giorgiosironi if you would like to review, at minimum you could let me know if the variable naming and having multiple return types for parse() are ok (this is just one way I thought for backwards compatibility). You can also dig into ElementTree as I did to see how the doctype hooks are implemented. This seems to work with my tests anyway.

File: elifetools/xmlio.py [hunk 1]
```diff
@@ -7,21 +7,36 @@
...
     """
     Register namespaces globally
     """
     ElementTree.register_namespace("mml","http://www.w3.org/1998/Math/MathML")
     ElementTree.register_namespace("xlink","http://www.w3.org/1999/xlink")
+    ElementTree.register_namespace("ali","http://www.niso.org/schemas/ali/1.0/")
 
-def parse(filename):
-    parser = ElementTree.XMLParser(html=0, target=None, encoding='utf-8')
+def parse(filename, return_doctype_dict=False):
```
Comment (giorgiosironi) [inline elifetools/xmlio.py]: from the outside world this looks backwards compatible, it is always using `CustomXMLParser` (also in the backward compatible case) but it seems it doesn't change the behavior

File: elifetools/xmlio.py [hunk 2]
```diff
@@ -7,21 +7,36 @@
...
+    parser = CustomXMLParser(html=0, target=None, encoding='utf-8')
 
     tree = ElementTree.parse(filename, parser)
     root = tree.getroot()
-    
-    return root
-    
+
+    if return_doctype_dict is True:
+        return root, parser.doctype_dict
```
Comment (giorgiosironi) [inline elifetools/xmlio.py]: also the multiple return types look correct, preserving existing callers which expect a single value

File: elifetools/xmlio.py [hunk 3]
```diff
@@ -7,21 +7,36 @@
 xmlio can do input and output of XML, allowing it to be edited using ElementTree library
 """
 
+class CustomXMLParser(ElementTree.XMLParser):
+    doctype_dict = {}
```
Comment (giorgiosironi) [inline elifetools/xmlio.py]: the scope of these values is limited to the parser object, which is vastly better than the main Python module

**Comment by gnott** [comment]
Thanks @giorgiosironi I see no suggestions for changes if I read correctly. I will merge this and then do the elife-bot side of things.

**Pull request merged**